include/zephyr/posix/sys/utsname.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (c) 2023 Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+#define ZEPHYR_INCLUDE_POSIX_SYS_UTSNAME_H_
+
+#include <errno.h>
+#include <zephyr/toolchain/common.h>
+#include <zephyr/net/hostname.h>
+#include "version.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* sysname */
+#define SYS_NAMELEN     sizeof("Zephyr")
+
+/* nodename */
+#ifdef CONFIG_NET_HOSTNAME_ENABLE
+#define NODE_NAMELEN    NET_HOSTNAME_MAX_LEN
+#else
+#define NODE_NAMELEN    SYS_NAMELEN
+#endif
+
+/* release */
+#define RELEASE_NAMELEN sizeof(KERNEL_VERSION_STRING)
+
+/* version */
+#if defined(__DATE__) && defined(__TIME__)
+#define DATETIME_LEN (sizeof(__DATE__) + sizeof(__TIME__))
+#else
+#define DATETIME_LEN 0
+#endif
+
+#ifdef BUILD_VERSION
+#define VERSION_BUILD STRINGIFY(BUILD_VERSION)
+#else
+#define VERSION_BUILD KERNEL_VERSION_STRING
+#endif
+#define VERSION_BUILD_LEN sizeof(VERSION_BUILD)
+
+#define VERSION_NAMELEN (VERSION_BUILD_LEN + DATETIME_LEN)
+
+/* machine */
+#define MACHINE_NAMELEN sizeof(CONFIG_ARCH)
+
+struct utsname {
```

Comment:
Minor comments - it looks as though the array length should be unspecified according to the spec, but if that would cause errors, then no problem.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_utsname.h.html